sockjsclient/xhr.go

```diff
@@ -0,0 +1,183 @@
+package sockjsclient
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/cookiejar"
+	"sync"
+)
+
+// the implementation of New() doesn't have any error to be returned yet it
+// returns, so it's totally safe to neglect the error
+var cookieJar, _ = cookiejar.New(nil)
+
+type XHRSession struct {
+	mu sync.Mutex
+
+	client     *http.Client
+	sessionURL string
+	sessionID  string
+	messages   []string
+	opened     bool
+}
+
+// NewXHRSession returns a new XHRSession, a SockJS client which supports
+// xhr-polling
+// http://sockjs.github.io/sockjs-protocol/sockjs-protocol-0.3.3.html#section-74
+func NewXHRSession(opts *DialOptions) (*XHRSession, error) {
+	client := &http.Client{
+		// never make it less than the heartbeat delay from the sockjs server.
+		// If this is los, your requests to the server will time out, so you'll
+		// never receive the heartbeat frames.
+		Timeout: opts.Timeout,
+		// add this so we can make use of load balancer's sticky session features,
+		// such as AWS ELB
+		Jar: cookieJar,
+	}
+
+	// following /server_id/session_id should always be the same for every session
+	serverID := threeDigits()
+	sessionID := randomStringLength(20)
+	sessionURL := opts.BaseURL + "/" + serverID + "/" + sessionID
+
+	// start the initial session handshake
+	sessionResp, err := client.Post(sessionURL+"/xhr", "text/plain", nil)
+	if err != nil {
+		return nil, err
+	}
+	defer sessionResp.Body.Close()
+
+	buf := bufio.NewReader(sessionResp.Body)
+	frame, err := buf.ReadByte()
+	if err != nil {
+		return nil, err
+	}
+
+	if frame != 'o' {
+		return nil, fmt.Errorf("can't start session, invalid frame: %s", frame)
+	}
+
+	return &XHRSession{
+		client:     client,
+		sessionID:  sessionID,
+		sessionURL: sessionURL,
+		opened:     true,
+	}, nil
+}
+
+func (x *XHRSession) ID() string {
+	return x.sessionID
+}
+
+func (x *XHRSession) Recv() (string, error) {
+	// Return previously received messages if there is any.
+	if len(x.messages) > 0 {
+		msg := x.messages[0]
+		x.messages = x.messages[1:]
+		return msg, nil
+	}
+
+	// start to poll from the server until we receive something
+	for {
+		resp, err := x.client.Post(x.sessionURL+"/xhr", "text/plain", nil)
+		if err != nil {
+			return "", err
+		}
+		defer resp.Body.Close()
+
+		buf := bufio.NewReader(resp.Body)
+
+		// returns an error if buffer is empty
+		frame, err := buf.ReadByte()
+		if err != nil {
+			return "", err
+		}
+
+		switch frame {
+		case 'o':
+			// session started
+			x.mu.Lock()
+			x.opened = true
+			x.mu.Unlock()
+			continue
+		case 'a':
+			// received an array of messages
+			data, err := ioutil.ReadAll(buf)
+			if err != nil {
+				return "", err
+			}
+
```

Comment:
you can Decode all the messages with 
```
err = json.NewDecoder(buf).Decode(&message)
```
and omit the ioutill.ReadAll & json.Unmarshal